src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,146 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
-        RAY_LOG(DEBUG)
-            << "Repeated reconnection in "
-            << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
-            << " milliseconds, return directly.";
-        return;
-      }
-
-      RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
-                     << address.second;
-      if (Ping(address.first, address.second, 100)) {
-        // If `last_reconnect_address_` port is -1, it means that this is the first
-        // connection and no log will be printed.
-        if (last_reconnect_address_.second != -1) {
-          RAY_LOG(INFO) << "Reconnected to GCS server: " << address.first << ":"
-                        << address.second;
-        }
-        reconnected = true;
-        break;
-      }
-    }
-    std::this_thread::sleep_for(
-        std::chrono::milliseconds(kGCSReconnectionRetryIntervalMs));
+
+    RAY_LOG(DEBUG) << "Attemptting to reconnect to GCS server: " << address.first << ":"
+                   << address.second;
+    gcs_rpc_client_->ResetHeartBeatInfoClient(address.first, address.second,
+                                              *client_call_manager_);
+    rpc::CheckAliveRequest request;
+    gcs_rpc_client_->CheckAlive(
+        request,
+        [this, start, address](const Status &status, const rpc::CheckAliveReply &reply) {
+          RAY_LOG(DEBUG) << "CheckAlive returns " << status;
+          OnReconnectionFinished(status, start, address);
+        });
+  } else {
+    RAY_LOG(WARNING) << "Can not get gcs server address.";
+    ExecuteReconnectedCallbacks();
+    reconnect_in_progress_ = false;
   }
+}
+
+void GcsClient::OnReconnectionFinished(const Status &status, absl::Time start,
+                                       std::pair<std::string, int> address) {
+  absl::MutexLock l(&reconnect_flag_mutex_);
+  if (status.ok()) {
+    RAY_LOG(INFO) << "Reconnected to GCS server: " << address.first << ":"
+                  << address.second;
 
-  if (reconnected) {
     gcs_rpc_client_->Reset(address.first, address.second, *client_call_manager_);
     last_reconnect_address_ = address;
     last_reconnect_timestamp_ms_ = current_sys_time_ms();
+    ExecuteReconnectedCallbacks();
+    reconnect_in_progress_ = false;
   } else {
-    RAY_LOG(FATAL) << "Couldn't reconnect to GCS server. The last attempted GCS "
-                      "server address was "
-                   << address.first << ":" << address.second;
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(kGCSReconnectionRetryIntervalMs));
+    if (absl::Now() - start <
+        absl::Seconds(RayConfig::instance().gcs_rpc_server_reconnect_timeout_s())) {
+      DoReconnect(start);
+    } else {
+      RAY_LOG(FATAL) << "Couldn't reconnect to GCS server. The last attempted GCS "
+                        "server address was "
+                     << address.first << ":" << address.second;
+      reconnect_in_progress_ = false;
```

Comment:
Not necessary since RAY_LOG(FATAL) will crash the process

Reply:
Legacy code too.😂 RAY_LOG(WARNING) instead.